grpc/src/main/java/com/linecorp/armeria/common/grpc/GoogleGrpcStatusFunction.java

```diff
@@ -0,0 +1,79 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.grpc;
+
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import io.grpc.Status;
+import io.grpc.StatusException;
+import io.grpc.StatusRuntimeException;
+import io.grpc.protobuf.lite.ProtoLiteUtils;
+
+/**
+ * A {@link GrpcStatusFunction} that provides a way to include details of a status into a {@link Metadata}.
+ * The details are stored in the `grpc-status-details-bin` key in the {@link Metadata}
+ * as a {@link com.google.rpc.Status}.
+ * You can implement a mapping function to converts {@link Throwable} into a {@link com.google.rpc.Status}
+ * which is stored in the `grpc-status-details-bin` key in the {@link Metadata}.
+ * If a given {@link Throwable} is an instance of either {@link StatusRuntimeException} or
+ * {@link StatusException}, returns a {@link Status} retrieved from the exception
+ * prior to the mapping function you implemented.
+ */
+@UnstableApi
+public interface GoogleGrpcStatusFunction extends GrpcStatusFunction {
+
+    Key<com.google.rpc.Status> GRPC_STATUS_DETAILS_BIN_KEY = Key.of(
+            "grpc-status-details-bin",
+            ProtoLiteUtils.metadataMarshaller(com.google.rpc.Status.getDefaultInstance()));
+
+    @Nullable
+    @Override
+    default Status apply(RequestContext ctx, Throwable throwable, Metadata metadata) {
+        final Throwable cause = Exceptions.peel(requireNonNull(throwable, "throwable"));
+        if (cause instanceof StatusRuntimeException) {
+            return ((StatusRuntimeException) cause).getStatus();
+        }
+        if (cause instanceof StatusException) {
+            return ((StatusException) cause).getStatus();
+        }
+        final com.google.rpc.Status statusProto = applyStatusProto(ctx, cause, metadata);
+        if (statusProto == null) {
+            return null;
+        }
+        final Status status = Status.fromCodeValue(statusProto.getCode())
+                                    .withDescription(statusProto.getMessage());
+        metadata.discardAll(GRPC_STATUS_DETAILS_BIN_KEY);
+        metadata.put(GRPC_STATUS_DETAILS_BIN_KEY, statusProto);
+        return status;
+    }
+
+    /**
+     * Maps the specified {@link Throwable} to a {@link com.google.rpc.Status},
+     * and mutates the specified {@link Metadata}.
+     * The `grpc-status-details-bin` key is ignored since it will be overwritten
+     * by {@link GoogleGrpcStatusFunction#apply(RequestContext, Throwable, Metadata)}.
+     * If {@code null} is returned, the built-in mapping rule is used by default.
+     */
+    com.google.rpc.Status applyStatusProto(RequestContext ctx, Throwable throwable, Metadata metadata);
```

Comment:
nit: `@Nullable`?

Reply:
Can we also make this method as default? I'm asking because, if we can make it, we can provide a default implementation for this interface so that users easily use this function.
`GrpcService.builder().exceptionMapping(GoogleGrpcStatusFunction.of())`

Can we do something like below? 🤔  
```java
default com.google.rpc.Status applyStatusProto(RequestContext ctx, Throwable throwable, Metadata metadata) {
    final Status status = Status.fromThrowable(throwable);
    return com.google.rpc.Status.newBuilder()
                                .setCode(status.getCode().value())
                                .setMessage(throwable.getMessage())
                                .build();
}
```

Reply:
It seems good!

Reply:
I found a problem with this deafult implmentation.
We already have checked `throwable` to allow users to throw `StatusRuntimeException` created from `StatusProto.toStatusRuntimeException()` in users' own code below.

https://github.com/line/armeria/blob/6b3b73193eee8e4c131366ebf7ad4a946e38b459/grpc/src/main/java/com/linecorp/armeria/common/grpc/GoogleGrpcStatusFunction.java#L53-L59

Therefore, The default implementation you suggested would not be used.

Reply:
So, you mean adding the `throwable.getMessage()` to `com.google.rpc.Status` wouldn't be useful, right?
Then, let's go without the default implementation. 😉